File: backend_java/src/main/java/com/cadence/api/export/ExportWriter.java
package com.cadence.api.export;

import com.cadence.api.activities.Activity;
import com.cadence.api.activities.ActivityRepository;
import com.cadence.api.activities.ActivityService;
import com.cadence.api.activities.LapMapper;
import com.cadence.api.activities.LapRepository;
import com.cadence.api.activities.StreamService;
import com.cadence.api.activities.dto.LapResponse;
import com.cadence.api.common.domain.Sport;
import com.cadence.api.export.dto.ActivityExportEntry;
import com.cadence.api.export.dto.ExportCounts;
import com.cadence.api.gear.Bike;
import com.cadence.api.gear.BikeRepository;
import com.cadence.api.gear.ComponentRepository;
import com.cadence.api.gear.GearMapper;
import com.cadence.api.gear.GearService;
import com.cadence.api.gear.ShoeRepository;
import com.cadence.api.gear.ShoeService;
import com.cadence.api.gear.dto.BikeResponse;
import com.cadence.api.gear.dto.ComponentResponse;
import com.cadence.api.gear.dto.ShoeResponse;
import com.cadence.api.races.RaceRepository;
import com.cadence.api.races.RaceService;
import com.cadence.api.scheduling.ScheduledWorkoutRepository;
import com.cadence.api.scheduling.SchedulingMapper;
import com.cadence.api.workouts.Workout;
import com.cadence.api.workouts.WorkoutMapper;
import com.cadence.api.workouts.WorkoutRepository;
import com.cadence.api.workouts.dto.WorkoutDetailResponse;
import jakarta.persistence.EntityManager;
import java.time.Instant;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.IntConsumer;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;
import tools.jackson.core.JsonGenerator;

/**
 * Streams one athlete's full data export straight into the given generator, one section/record
 * at a time, so peak memory stays bounded to a single activity's laps+streams rather than the
 * athlete's whole history. Read-only and transactional so lazy associations (e.g. Shoe's model/
 * version, walked by ShoeService.toResponse) stay resolvable for the whole call - open-in-view is
 * off, so without this boundary those would throw once touched outside a session.
 *
 * <p>{@code entityManager.clear()} after every activity is load-bearing, not cosmetic: Hibernate's
 * persistence context pins every entity it loads for the life of the transaction, so without this
 * a large account (thousands of activities, millions of {@code Record} rows) fills the identity
 * map and OOMs well before the JSON/gzip side of this class ever sees that much data. Detaching
 * per-activity is safe because every field touched on an already-processed (or not-yet-processed)
 * Activity afterward is either a plain column or a lazy association read via {@code .getId()} only
 * (see ActivityService.toResponse / SchedulingMapper), which Hibernate resolves off the proxy
 * itself without needing the owning session.
 */
@Component
public class ExportWriter {

	// Persisting after every single item would mean one DB write per row in a multi-thousand-
	// activity export - throttle to every 20 items (still gives a smooth-looking bar).
	private static final int PROGRESS_INTERVAL = 20;

	/** Wraps an onProgress callback with throttling. {@link #flush()} bypasses the throttle -
	 * called at section boundaries so the persisted count never drifts far behind reality
	 * (otherwise a section with, say, 7 items never fires at all, since 7 % 20 != 0). */
	private static final class Progress {
		private final IntConsumer onProgress;
		private int count;

		Progress(IntConsumer onProgress) {
			this.onProgress = onProgress;
		}

		void tick() {
			count++;
			if (count % PROGRESS_INTERVAL == 0) {
				onProgress.accept(count);
			}
		}

		void tick(int n) {
			for (int i = 0; i < n; i++) {
				tick();
			}
		}

		void flush() {
			onProgress.accept(count);
		}
	}

	private final ActivityRepository activityRepository;
	private final ActivityService activityService;
	private final LapRepository lapRepository;
	private final LapMapper lapMapper;
	private final StreamService streamService;
	private final RaceRepository raceRepository;
	private final RaceService raceService;
	private final WorkoutRepository workoutRepository;
	private final WorkoutMapper workoutMapper;
	private final ScheduledWorkoutRepository scheduledWorkoutRepository;
	private final SchedulingMapper schedulingMapper;
	private final GearService gearService;
	private final ShoeService shoeService;
	private final BikeRepository bikeRepository;
	private final ShoeRepository shoeRepository;
	private final ComponentRepository componentRepository;
	private final GearMapper gearMapper;
	private final EntityManager entityManager;

	public ExportWriter(ActivityRepository activityRepository, ActivityService activityService, LapRepository lapRepository,
			LapMapper lapMapper, StreamService streamService, RaceRepository raceRepository, RaceService raceService,
			WorkoutRepository workoutRepository, WorkoutMapper workoutMapper,
			ScheduledWorkoutRepository scheduledWorkoutRepository, SchedulingMapper schedulingMapper,
			GearService gearService, ShoeService shoeService, BikeRepository bikeRepository, ShoeRepository shoeRepository,
			ComponentRepository componentRepository, GearMapper gearMapper, EntityManager entityManager) {
		this.activityRepository = activityRepository;
		this.activityService = activityService;
		this.lapRepository = lapRepository;
		this.lapMapper = lapMapper;
		this.streamService = streamService;
		this.raceRepository = raceRepository;
		this.raceService = raceService;
		this.workoutRepository = workoutRepository;
		this.workoutMapper = workoutMapper;
		this.scheduledWorkoutRepository = scheduledWorkoutRepository;
		this.schedulingMapper = schedulingMapper;
		this.gearService = gearService;
		this.shoeService = shoeService;
		this.bikeRepository = bikeRepository;
		this.shoeRepository = shoeRepository;
		this.componentRepository = componentRepository;
		this.gearMapper = gearMapper;
		this.entityManager = entityManager;
	}

	/** {@code sportFilter} narrows activities/races/workouts/scheduled_workouts to one discipline;
	 * {@code null} exports everything, unchanged from before this filter existed. Equipment is
	 * always exported in full either way - it's small, and shoes have no sport of their own to
	 * filter by.
	 *
	 * <p>Annotated independently rather than just delegating to the 4-arg overload below: a
	 * same-class call to that overload wouldn't go through this bean's transactional proxy
	 * (Spring AOP self-invocation), silently dropping {@code @Transactional} and breaking every
	 * lazy association this class relies on being resolvable - caught by
	 * ExportWriterIntegrationTest failing with LazyInitializationException the first time this
	 * was missed. */
	@Transactional(readOnly = true)
	public void write(String athleteId, Sport sportFilter, JsonGenerator generator) {
		write(athleteId, sportFilter, generator, step -> { });
	}

	/** As {@link #write(String, Sport, JsonGenerator)}, but calls {@code onStep} with one of
	 * "equipment"/"workouts"/"activities"/"races"/"scheduled_workouts" right as each section
	 * starts - what a progress dialog polls ExportJob.currentStep for. {@code onStep} is expected
	 * to persist independently of this method's own transaction (see ExportProgressUpdater) - it
	 * must not be assumed to see the writes this transaction hasn't committed yet, or vice versa. */
	@Transactional(readOnly = true)
	public void write(String athleteId, Sport sportFilter, JsonGenerator generator, Consumer<String> onStep) {
		write(athleteId, sportFilter, generator, onStep, total -> { }, processed -> { });
	}

	/** As {@link #write(String, Sport, JsonGenerator, Consumer)}, but also turns onStep's coarse
	 * per-section signal into a fine-grained item-level progress bar *scoped to the current
	 * section* - e.g. while onStep is "activities", onTotal/onProgress report activities
	 * specifically, not a blend across every kind of data. {@code onTotal} fires once per
	 * section, right after onStep, with that section's own count from computeCounts (the same
	 * cheap count queries the "counts" file field already computes - so this is free, no second
	 * query); {@code onProgress} fires with a running count *reset to 0 at the start of each
	 * section* as each item in it is written, throttled (see Progress) so a multi-thousand-row
	 * export doesn't turn into a DB write per row. "equipment" covers bikes+shoes+components
	 * combined (one section/one onStep call), so its total is the sum of those three. Same
	 * persist-independently contract as onStep. */
	@Transactional(readOnly = true)
	public void write(String athleteId, Sport sportFilter, JsonGenerator generator, Consumer<String> onStep,
			IntConsumer onTotal, IntConsumer onProgress) {
		generator.writeStartObject();
		generator.writeStringProperty("generated_at", Instant.now().toString());
		generator.writeStringProperty("athlete_id", athleteId);
		ExportCounts counts = computeCounts(athleteId, sportFilter);
		generator.writePOJOProperty("counts", counts);

		// Equipment and workouts come before activities (which reference bike/shoe/workout ids),
		// and activities come before races/scheduled_workouts (which reference activity ids) -
		// dependency order, so a forward-only streaming importer can resolve id references in one
		// pass without having to look ahead. See ImportReader for the reader side of this contract.
		Progress progress = startSection(onStep, onTotal, onProgress, "equipment",
				(int) (counts.bikes() + counts.shoes() + counts.components()));
		writeEquipment(athleteId, generator, progress);
		progress.flush();

		progress = startSection(onStep, onTotal, onProgress, "workouts", (int) counts.workouts());
		writeWorkouts(athleteId, sportFilter, generator, progress);
		progress.flush();

		progress = startSection(onStep, onTotal, onProgress, "activities", (int) counts.activities());
		writeActivities(athleteId, sportFilter, generator, progress);
		progress.flush();

		progress = startSection(onStep, onTotal, onProgress, "races", (int) counts.races());
		writeRaces(athleteId, sportFilter, generator, progress);
		progress.flush();

		progress = startSection(onStep, onTotal, onProgress, "scheduled_workouts", (int) counts.scheduledWorkouts());
		writeScheduledWorkouts(athleteId, sportFilter, generator, progress);
		progress.flush();

		generator.writeEndObject();
	}

	private static Progress startSection(
			Consumer<String> onStep, IntConsumer onTotal, IntConsumer onProgress, String step, int total) {
		onStep.accept(step);
		onTotal.accept(total);
		return new Progress(onProgress);
	}

	/** Cheap upfront count queries (no row materialization) for the "counts" metadata block at
	 * the top of the file - the same filtered scope each writeX method below uses. */
	private ExportCounts computeCounts(String athleteId, Sport sportFilter) {
		long activities = sportFilter == null ? activityRepository.countByAthleteId(athleteId)
				: activityRepository.countByAthleteIdAndSport(athleteId, sportFilter);
		long races = sportFilter == null ? raceRepository.countByAthleteId(athleteId)
				: raceRepository.countByAthleteIdAndEffectiveSport(athleteId, sportFilter);
		long workouts = sportFilter == null ? workoutRepository.countByCreatedById(athleteId)
				: workoutRepository.countByCreatedByIdAndSport(athleteId, sportFilter);
		long scheduledWorkouts = sportFilter == null ? scheduledWorkoutRepository.countByAthleteId(athleteId)
				: scheduledWorkoutRepository.countByAthleteIdAndWorkoutSport(athleteId, sportFilter);
		// Equipment is always exported in full, regardless of the sport filter - it's small, and
		// shoes have no sport of their own to filter by.
		return new ExportCounts(activities, races, workouts, scheduledWorkouts, bikeRepository.countByAthleteId(athleteId),
				shoeRepository.countByAthleteIdAndRetiredFalse(athleteId), componentRepository.countByBikeAthleteId(athleteId));
	}

	private void writeActivities(String athleteId, Sport sportFilter, JsonGenerator generator, Progress progress) {
		// A multisport parent's own sport is MULTISPORT, not e.g. BIKE, so filtering by sport here
		// naturally excludes multisport parents while still including any matching individual legs -
		// no special-case code needed. A leg whose parent got excluded this way just carries a
		// parent_activity_id ImportReader won't find in this file; its deferred-link resolution
		// already handles that (skips the link rather than failing).
		List<Activity> activities = sportFilter == null
				? activityRepository.findByAthleteIdOrderByStartDate(athleteId)
				: activityRepository.findByAthleteIdAndSportOrderByStartDate(athleteId, sportFilter);

		generator.writeArrayPropertyStart("activities");
		for (Activity activity : activities) {
			List<LapResponse> laps = lapRepository.findByActivityIdOrderByIndex(activity.getId()).stream()
					.map(lapMapper::toResponse)
					.toList();
			// "high" resolution == no downsampling (StreamService.getStreams step=1 default) - a
			// full data export shouldn't quietly thin out the athlete's own recorded data.
			generator.writePOJO(new ActivityExportEntry(
					activityService.toResponse(activity).withoutTransientThresholdState(), laps,
					streamService.getStreams(activity, null, "high")));
			entityManager.clear();
			progress.tick();
		}
		generator.writeEndArray();
	}

	private void writeRaces(String athleteId, Sport sportFilter, JsonGenerator generator, Progress progress) {
		generator.writeArrayPropertyStart("races");
		for (var race : raceRepository.findByAthleteIdOrderByDateAsc(athleteId)) {
			// A race's own sport can be null even when it's linked to an activity - linking to
			// an *existing* race (as opposed to "mark this activity as a race", which sets sport
			// at creation) never backfills it. Fall back to the linked activity's sport so a
			// sport-scoped export doesn't silently drop these. A race with no sport recorded and
			// no linked activity to infer one from is still excluded under a filter - ambiguous
			// otherwise.
			Sport effectiveSport = race.getSport() != null ? race.getSport()
					: race.getActivity() != null ? race.getActivity().getSport() : null;
			if (sportFilter == null || sportFilter.equals(effectiveSport)) {
				generator.writePOJO(raceService.toResponse(race));
				progress.tick();
			}
		}
		generator.writeEndArray();
	}

	private void writeWorkouts(String athleteId, Sport sportFilter, JsonGenerator generator, Progress progress) {
		generator.writeArrayPropertyStart("workouts");
		for (Workout workout : workoutRepository.findByCreatedByIdOrderByIdDesc(athleteId)) {
			if (sportFilter == null || sportFilter.equals(workout.getSport())) {
				generator.writePOJO(new WorkoutDetailResponse(
						workoutMapper.toResponse(workout), workoutMapper.toStepTree(workout.getSteps())));
				progress.tick();
			}
		}
		generator.writeEndArray();
	}

	private void writeScheduledWorkouts(String athleteId, Sport sportFilter, JsonGenerator generator, Progress progress) {
		generator.writeArrayPropertyStart("scheduled_workouts");
		for (var scheduled : scheduledWorkoutRepository.findByAthleteIdOrderByDate(athleteId)) {
			if (sportFilter == null || sportFilter.equals(scheduled.getWorkout().getSport())) {
				generator.writePOJO(schedulingMapper.toResponse(scheduled));
				progress.tick();
			}
		}
		generator.writeEndArray();
	}

	private void writeEquipment(String athleteId, JsonGenerator generator, Progress progress) {
		List<Bike> bikes = gearService.listBikes(athleteId);

		generator.writeObjectPropertyStart("equipment");

		List<BikeResponse> bikeResponses = bikes.stream().map(gearService::toBikeResponse).toList();
		generator.writePOJOProperty("bikes", bikeResponses);
		progress.tick(bikeResponses.size());

		// Shoe/Component have no retired flag of their own to filter here; listShoes already
		// excludes retired shoes (findByAthleteIdAndRetiredFalseOrderByIdDesc) - matches the
		// "active gear only" scope for this export.
		List<ShoeResponse> shoeResponses = shoeService.listShoes(athleteId).stream().map(shoeService::toResponse).toList();
		generator.writePOJOProperty("shoes", shoeResponses);
		progress.tick(shoeResponses.size());

		List<ComponentResponse> componentResponses = bikes.stream()
				.flatMap(bike -> componentRepository.findByBikeId(bike.getId()).stream())
				.map(gearMapper::toResponse)
				.toList();
		generator.writePOJOProperty("components", componentResponses);
		progress.tick(componentResponses.size());

		generator.writeEndObject();
	}
}

File: backend/dataexport/export_writer.py
"""Streams one athlete's full data export to a gzip file on local disk.

The JSON shape here is a deliberate mirror of the Java backend's ExportWriter
(backend_java/src/main/java/com/cadence/api/export/ExportWriter.java) - same top-level
keys, same nesting, same field names - so a file exported from either backend can be
imported into the other. See import_reader.py for the read side of that contract.

Unlike Java (which had to guard against Hibernate's persistence context pinning every
loaded entity in memory - see ExportWriter's Javadoc), Django's ORM has no equivalent
identity map, so there's no analogous per-activity "clear the session" step needed here.
The one thing this still has to get right is not materializing a huge activity's whole
Record queryset at once - see _build_streams' use of .iterator().
"""

import gzip
import json
import os
from collections.abc import Callable
from typing import Any

from django.core.files.storage import default_storage
from django.core.serializers.json import DjangoJSONEncoder
from django.db.models import Q
from django.utils import timezone

from activities.models import Activity, Lap, Record
from activities.serializers import ActivitySerializer, LapSerializer
from gear.models import Bike, Component, Shoe
from gear.serializers import BikeSerializer, ComponentSerializer, ShoeSerializer
from races.models import Race
from races.serializers import RaceSerializer
from scheduling.models import ScheduledWorkout
from scheduling.serializers import ScheduledWorkoutSerializer
from workouts.models import Workout
from workouts.serializers import WorkoutDetailSerializer

OnStep = Callable[[str], None]
OnTotal = Callable[[int], None]
OnProgress = Callable[[int], None]

# Persisting after every single item would mean one DB write per row in a multi-thousand-
# activity export - throttle to every 20 items (still gives a smooth-looking bar).
_PROGRESS_INTERVAL = 20


class _Progress:
    """Wraps an `on_progress` callback with throttling. `flush()` bypasses the throttle -
    called at section boundaries so the persisted count never drifts far behind reality
    (otherwise a section with, say, 7 items never fires at all, since 7 % 20 != 0)."""

    def __init__(self, on_progress: OnProgress | None):
        self._on_progress = on_progress
        self.count = 0

    def tick(self) -> None:
        self.count += 1
        if self._on_progress and self.count % _PROGRESS_INTERVAL == 0:
            self._on_progress(self.count)

    def flush(self) -> None:
        if self._on_progress:
            self._on_progress(self.count)


# Same 12 scalar fields (in the same order) as the Java backend's StreamService.SCALAR_FIELDS -
# always present, filled with None wherever the source data lacks that channel.
STREAM_FIELDS = [
    "time",
    "power",
    "heartrate",
    "cadence",
    "altitude",
    "distance",
    "speed",
    "air_temp",
    "humidity",
    "core_temp",
    "skin_temp",
    "heat_strain",
]
_RECORD_ATTR = {
    "time": "t",
    "power": "power",
    "heartrate": "heartrate",
    "cadence": "cadence",
    "altitude": "altitude",
    "distance": "distance_km",
    "speed": "speed",
    "air_temp": "air_temp",
    "humidity": "humidity",
    "core_temp": "core_temp",
    "skin_temp": "skin_temp",
    "heat_strain": "heat_strain",
}


def _dump(value: Any) -> bytes:
    return json.dumps(value, cls=DjangoJSONEncoder).encode("utf-8")


def _write_array(gz: gzip.GzipFile, items: Any, progress: "_Progress | None" = None) -> None:
    gz.write(b"[")
    first = True
    for item in items:
        if not first:
            gz.write(b",")
        first = False
        gz.write(_dump(item))
        if progress:
            progress.tick()
    gz.write(b"]")


def _build_streams(activity: Activity) -> dict:
    fields: dict[str, list] = {name: [] for name in STREAM_FIELDS}
    latlng: list[list[float]] = []
    for record in Record.objects.filter(activity=activity).iterator(chunk_size=2000):
        for field_name, attr in _RECORD_ATTR.items():
            fields[field_name].append(getattr(record, attr))
        if activity.has_gps and record.lat is not None and record.lng is not None:
            latlng.append([record.lat, record.lng])
    if activity.has_gps:
        fields["latlng"] = latlng
    return {"object": "streams", "resolution": "high", "fields": fields}


def _bikes_qs(athlete_id: str):
    return Bike.objects.filter(athlete_id=athlete_id)


def _shoes_qs(athlete_id: str):
    # Active gear only, matching the Java backend's export scope.
    return Shoe.objects.filter(athlete_id=athlete_id, retired=False)


def _components_qs(athlete_id: str):
    return Component.objects.filter(bike__athlete_id=athlete_id)


def _workouts_qs(athlete_id: str, sport: str | None):
    qs = Workout.objects.filter(created_by_id=athlete_id)
    if sport:
        qs = qs.filter(sport=sport)
    return qs


def _activities_qs(athlete_id: str, sport: str | None):
    # A multisport parent's own sport is "multisport", not e.g. "bike", so filtering here
    # naturally excludes multisport parents while still including matching individual legs -
    # no special-case code needed (same reasoning as the Java backend's ExportWriter).
    qs = Activity.objects.filter(athlete_id=athlete_id).order_by("start_date")
    if sport:
        qs = qs.filter(sport=sport)
    return qs


def _races_qs(athlete_id: str, sport: str | None):
    qs = Race.objects.filter(athlete_id=athlete_id).select_related("activity").order_by("date")
    if sport:
        # A race's own sport can be blank even when it's linked to an activity - linking to
        # an *existing* race (as opposed to "mark this activity as a race", which sets sport
        # at creation) never backfills it. Fall back to the linked activity's sport so a
        # sport-scoped export doesn't silently drop these. A race with no sport recorded and
        # no linked activity to infer one from is still excluded under a filter - ambiguous
        # otherwise (matches the Java backend).
        qs = qs.filter(Q(sport=sport) | Q(sport="", activity__sport=sport))
    return qs


def _scheduled_workouts_qs(athlete_id: str, sport: str | None):
    qs = ScheduledWorkout.objects.filter(athlete_id=athlete_id).order_by("date")
    if sport:
        qs = qs.filter(workout__sport=sport)
    return qs


def _counts(athlete_id: str, sport: str | None) -> dict:
    """Cheap upfront .count() queries (no row materialization) for the "counts" metadata block
    at the top of the file - the same filtered scope each _write_X function below uses, built
    from the same querysets so the two can never drift apart."""
    return {
        "activities": _activities_qs(athlete_id, sport).count(),
        "races": _races_qs(athlete_id, sport).count(),
        "workouts": _workouts_qs(athlete_id, sport).count(),
        "scheduled_workouts": _scheduled_workouts_qs(athlete_id, sport).count(),
        # Equipment is always exported in full, regardless of the sport filter - it's small,
        # and shoes have no sport of their own to filter by.
        "bikes": _bikes_qs(athlete_id).count(),
        "shoes": _shoes_qs(athlete_id).count(),
        "components": _components_qs(athlete_id).count(),
    }


def _write_equipment(gz: gzip.GzipFile, athlete_id: str, progress: "_Progress | None" = None) -> None:
    gz.write(b'{"bikes":')
    _write_array(gz, (BikeSerializer(b).data for b in _bikes_qs(athlete_id)), progress)

    gz.write(b',"shoes":')
    shoes = _shoes_qs(athlete_id).select_related("shoe_model_version__shoe_model")
    _write_array(gz, (ShoeSerializer(s).data for s in shoes), progress)

    gz.write(b',"components":')
    _write_array(gz, (ComponentSerializer(c).data for c in _components_qs(athlete_id)), progress)
    gz.write(b"}")


def _write_workouts(gz: gzip.GzipFile, athlete_id: str, sport: str | None, progress: "_Progress | None" = None) -> None:
    qs = _workouts_qs(athlete_id, sport).prefetch_related("steps")
    _write_array(gz, (WorkoutDetailSerializer(w).data for w in qs), progress)


# Deliberately never exported: transient, in-app-only bookkeeping (a pending threshold-increase
# suggestion, and whether detection has ever run) rather than data describing the activity
# itself - re-importing a file shouldn't resurrect an already-accepted-or-dismissed suggestion on
# a brand-new row, and a freshly-imported row should start out "not yet checked" like any other
# row import creates, not inherit whatever was true in the source account. ftp_snapshot/
# critical_run_power_snapshot/threshold_pace_snapshot ARE exported (ActivitySerializer already
# includes them) - those describe what actually happened, same as every other stat field.
_TRANSIENT_ACTIVITY_FIELDS = (
    "suggested_ftp",
    "suggested_critical_run_power",
    "suggested_threshold_pace",
    "threshold_checked",
)


def _write_activities(
    gz: gzip.GzipFile, athlete_id: str, sport: str | None, progress: "_Progress | None" = None
) -> None:
    gz.write(b"[")
    first = True
    for activity in _activities_qs(athlete_id, sport).iterator(chunk_size=200):
        if not first:
            gz.write(b",")
        first = False
        activity_data = ActivitySerializer(activity).data
        for field in _TRANSIENT_ACTIVITY_FIELDS:
            activity_data.pop(field, None)
        entry = {
            "activity": activity_data,
            "laps": LapSerializer(Lap.objects.filter(activity=activity), many=True).data,
            "streams": _build_streams(activity),
        }
        gz.write(_dump(entry))
        if progress:
            progress.tick()
    gz.write(b"]")


def _write_races(gz: gzip.GzipFile, athlete_id: str, sport: str | None, progress: "_Progress | None" = None) -> None:
    _write_array(gz, (RaceSerializer(r).data for r in _races_qs(athlete_id, sport)), progress)


def _write_scheduled_workouts(
    gz: gzip.GzipFile, athlete_id: str, sport: str | None, progress: "_Progress | None" = None
) -> None:
    _write_array(gz, (ScheduledWorkoutSerializer(s).data for s in _scheduled_workouts_qs(athlete_id, sport)), progress)


def _ensure_parent_dir(relative_path: str) -> str:
    full_path = default_storage.path(relative_path)
    os.makedirs(os.path.dirname(full_path), exist_ok=True)
    return full_path


def write_export(
    athlete_id: str,
    sport: str | None,
    relative_path: str,
    on_step: OnStep | None = None,
    on_total: OnTotal | None = None,
    on_progress: OnProgress | None = None,
) -> int:
    """Writes the gzip file to local disk (via default_storage's path resolution - this
    project's MEDIA_ROOT is local disk, which is what makes true incremental streaming
    writes possible; see the module docstring). Returns the file size in bytes.

    Field order matches Java's ExportWriter.write exactly: generated_at/athlete_id/counts
    first (counts is a handful of cheap .count() queries, computed upfront so a reader can see
    what the file contains without having to fully parse it), then equipment and workouts
    (which activities reference) before activities, activities (which races/scheduled_workouts
    reference) before those - dependency order, in case any consumer relies on it. `on_step`
    (if given) is called once per section, right as it starts. `on_total`/`on_progress` (if
    given) turn that into a fine-grained item-level progress bar *scoped to the current
    section* - e.g. while `on_step` is "activities", `on_total`/`on_progress` report "how many
    of this athlete's activities have been written", not a blended count across every kind of
    data. `on_total` fires once per section, right after `on_step`, with that section's own
    count from counts.py's .count() queries (the same ones the "counts" file field already
    computes - so this is free, no second query); `on_progress` fires with a running count
    *reset to 0 at the start of each section* as each item in it is written, throttled (see
    _Progress) so a multi-thousand-row export doesn't turn into a DB write per row. "equipment"
    covers bikes+shoes+components combined (they're one section/one on_step call), so its total
    is the sum of those three.
    """
    full_path = _ensure_parent_dir(relative_path)
    with gzip.GzipFile(full_path, mode="wb") as gz:
        gz.write(b'{"generated_at":')
        gz.write(_dump(timezone.now().isoformat().replace("+00:00", "Z")))
        gz.write(b',"athlete_id":')
        gz.write(_dump(athlete_id))
        counts = _counts(athlete_id, sport)
        gz.write(b',"counts":')
        gz.write(_dump(counts))

        def start_section(step: str, total: int) -> "_Progress":
            if on_step:
                on_step(step)
            if on_total:
                on_total(total)
            return _Progress(on_progress)

        progress = start_section("equipment", counts["bikes"] + counts["shoes"] + counts["components"])
        gz.write(b',"equipment":')
        _write_equipment(gz, athlete_id, progress)
        progress.flush()

        progress = start_section("workouts", counts["workouts"])
        gz.write(b',"workouts":')
        _write_workouts(gz, athlete_id, sport, progress)
        progress.flush()

        progress = start_section("activities", counts["activities"])
        gz.write(b',"activities":')
        _write_activities(gz, athlete_id, sport, progress)
        progress.flush()

        progress = start_section("races", counts["races"])
        gz.write(b',"races":')
        _write_races(gz, athlete_id, sport, progress)
        progress.flush()

        progress = start_section("scheduled_workouts", counts["scheduled_workouts"])
        gz.write(b',"scheduled_workouts":')
        _write_scheduled_workouts(gz, athlete_id, sport, progress)
        progress.flush()

        gz.write(b"}")
    return os.path.getsize(full_path)
